clang/lib/3C/CastPlacement.cpp

```diff
@@ -18,107 +18,244 @@
-    auto Fname = FD->getNameAsString();
-    if (!ConstraintResolver::canFunctionBeSkipped(Fname)) {
-      // Now we need to check the type of the arguments and corresponding
-      // parameters to see if any explicit casting is needed.
-      ProgramInfo::CallTypeParamBindingsT TypeVars;
-      if (Info.hasTypeParamBindings(CE, Context))
-        TypeVars = Info.getTypeParamBindings(CE, Context);
-      unsigned PIdx = 0;
-      for (const auto &A : CE->arguments()) {
-        if (PIdx < FD->getNumParams()) {
-          // Avoid adding incorrect casts to generic function arguments by
-          // removing implicit casts when on arguments with a consistently
-          // used generic type.
-          Expr *ArgExpr = A;
+    // Cast on arguments
+    unsigned PIdx = 0;
+    for (const auto &A : CE->arguments()) {
+      if (PIdx < FV->numParams() && (!FD || PIdx < FD->getNumParams())) {
+        // Avoid adding incorrect casts to generic function arguments by
+        // removing implicit casts when on arguments with a consistently
+        // used generic type.
+        Expr *ArgExpr = A;
+        if (FD) {
           const TypeVariableType *TyVar =
-              getTypeVariableType(FD->getParamDecl(PIdx));
+            getTypeVariableType(FD->getParamDecl(PIdx));
           if (TyVar && TypeVars.find(TyVar->GetIndex()) != TypeVars.end() &&
               TypeVars[TyVar->GetIndex()] != nullptr)
             ArgExpr = ArgExpr->IgnoreImpCasts();
+        }
 
-          CVarSet ArgumentConstraints = CR.getExprConstraintVars(ArgExpr);
-          ConstraintVariable *ParameterC = FV->getParamVar(PIdx);
-          for (auto *ArgumentC : ArgumentConstraints) {
-            if (needCasting(ArgumentC, ParameterC)) {
-              // We expect the cast string to end with "(".
-              std::string CastString = getCastString(ArgumentC, ParameterC);
-              surroundByCast(CastString, A);
-              break;
-            }
+        CVarSet ArgumentConstraints = CR.getExprConstraintVars(ArgExpr);
+        ConstraintVariable *ParamInt = FV->getInternalParam(PIdx);
+        ConstraintVariable *ParamExt = FV->getExternalParam(PIdx);
+        for (auto *ArgC : ArgumentConstraints) {
+          if (needCasting(ArgC, ArgC, ParamInt, ParamExt) != NO_CAST) {
+            surroundByCast(ArgC, ArgC, ParamInt, ParamExt, A);
+            break;
           }
         }
-        PIdx++;
+      }
+      PIdx++;
+    }
+
+    // Cast on return
+    if (Info.hasPersistentConstraints(CE, Context)) {
+      CVarSet ArgumentConstraints = CR.getExprConstraintVars(CE);
+      ConstraintVariable *RetInt = FV->getInternalReturn();
+      ConstraintVariable *RetExt = FV->getExternalReturn();
+      for (auto *ArgC : ArgumentConstraints) {
+        // Order of ParameterC and ArgumentC is reversed from when inserting
+        // parameter casts because assignment now goes from returned to its
+        // local use.
+        if (!Locator.exprHasCast(CE) &&
```

Comment:
If it has a cast already, you will not consider adding one. Is this because the cast will preclude the need to do that?